### PR TITLE
PLANET-5483: Add node container

### PR DESCRIPTION
**Pull request by lithrel**
Ref: https://jira.greenpeace.org/browse/PLANET-5483
Tries to do things for: https://jira.greenpeace.org/browse/PLANET-4752

> Investigate ways of pre-installing node/npm to our dev images

### Node container
- A basic [node container](https://hub.docker.com/_/node/) is used to run all npm operations (build/install/watch/etc.).  
- `node_modules` folders are excluded from mounting, which seems to help (MacOS) performances a bit.
- The container skips puppeteer Chromium download by default, to speed up first build.

### No Elastic for default build
Elastic container is not built by default. It's a step toward  [PLANET-4752](https://jira.greenpeace.org/browse/PLANET-4752)
The Elastic container is currently by far the most memory consuming of the stack, and memory limitation is a recurring problem during docker environment installs. It might be a configuration problem, but for a default/light version it's better to leave it out.  
Screenshot from 2021-02-08 12-58-03

### Variables changes
`PROJECT` and `DOCKER_COMPOSE_FILE`are replaced with their native counterpart `COMPOSE_PROJECT_NAME` and `COMPOSE_FILE`
Those are written in a `.env` file natively used by docker-compose; any command executed from this folder will use those values, which makes working with docker-compose a lot easier.  
Old variables can be removed from the docker-compose commands, and you don't have to run `eval $(make env)` in your shell sessions to work with it.  `COMPOSE_ENV` is dropped.
To change docker-compose env variables, edit `.env` file or run `COMPOSE_FILE=myfile.yml make env`.  

### Local scripts
Added a bunch of scripts, to replace some included in the docker repository, and some to split a bit more `make` operations.
```
scripts                 
├── build-assets.sh      // Build js/css assets
├── clean-content.sh     // Remove default content folder
├── clean-persistence.sh // Remove persistence folder
├── install-deps.sh      // Install composer and npm dependencies
├── update-repos.sh      // Update main theme and gutenberg blocks depos
└── watch.sh             // Run npm watch (options --theme-only --plugin-only)
```

### Update command
For devs not working on the main repos it can be easy to forget to get up to date with our changes.  
`make update` updates main repos and rebuilds assets.

### Removal of some hardcoded names
This (+ native variables) allows to keep different environments on the same machine just by building from different folders (well, folders with different names :grimacing: , or edit `COMPOSE_PROJECT_NAME` in `.env`)
Screenshot from 2021-02-03 16-33-35
They can't run simultaneously because of default port mapping, although editing ports in docker-compose file might work.

### Cosmetics & Help

Changed some `make status` subcommands so it displays passwords in a table and more links to tools in containers (Traefik, phpMyAdmin, ElasticHQ). 
Little modification of the `make help` function to parse exported variables.

## Todo next
- rsync operations take up to a minute and might not be needed after all (problem with ci?)
- replace npm with pnpm (or other?) for install/build, to try to speed up this step
- I think we can run dev env without openresty, by using php builtin webserver and a [handwritten router](https://stackoverflow.com/a/38926070). I didn't succeed at this, mostly because I don't seem to understand port mapping and declared networks.
- remove some error logs by maybe adding empty file `/app/source/public/wp-content/wflogs/rules.php`
- main repos (main theme and gutenberg plugin) are installed twice, by the container with composer-local, and by the Makefile here. Try to find a way to do this once only, either by building it properly in [there](https://github.com/greenpeace/planet4-docker/blob/develop/src/planet-4-151612/wordpress/etc/my_init.d/20_install_wordpress.sh) and removing it here, or making its installation there conditional (!dev env) and keeping it here (second option might make more sense because of local git config)

# Test

- Save your work
  - **Commit the things you're working on**, nothing ever goes as planned, it will probably disappear from your disk
- Clone this repository/branch in a different folder (with a **different folder name**)
  - ie: `git clone -b planet-5483-node-image --single-branch git@github.com:greenpeace/planet4-docker-compose.git p4dctest`
  - This version can be built in parallel of your current one **but it cannot run in parralel**, so `make stop` your current dev env.
  - get into the new folder
- For any retry - clean your system
  - clean planet4 things `make clean` (removing `defaultcontent` is optional)
  - (_very optional_) clean all docker things `docker system prune -a` (it will remove _everything_ docker has, including images previously downloaded and anything that doesn't run, for all projects)
- Check if you have a `.env` file - if you don't, everything's fine - if you do, please add `COMPOSE_FILE=docker-compose.yml` and `COMPOSE_PROJECT_NAME=p4dctest` if it's not in there
- You're on MacOS ? Restart docker
- Run `make dev`
  - run `docker-compose logs -ft` in another shell to check if some operations are still going on even if it crashes
- Be patient
- Report crashes

## Troubleshooting

- **Install times out after 10 minutes**
The process checking containers waited for too long. 
  - Check in the `php-fpm` logs if there is mention of a "Installation already underway, /app/source/.install exists. Sleeping...". If it is there, a crash happened; you cna try 2 things:
    - run `make dev` again, and while it waits, remove the lock file by running `docker-compose exec php-fpm sh -c "rm -f /app/source/.install"`
    - rerun the install by doing `make clean`  first (you don't have to remove defaultcontent)
  - Check in all logs for Out of Memory errors like "Command exited with non-zero status 137" or literally "Out of memory". Try to raise memory limit for docker if you can.
- **MacOS freezes when I run command x**
MacOS has a hard time handling file ownership mixed between your system and docker. Try running `make fix-ownership` and retry what you were doing. 

_`make dev` time, (without images/content/dependencies downloads, using cache)  
(downloads will significantly extend build time depending on your internet connection speed)_
| laptop | system | time (without downloads) |
| --- | --- | --- |
| x250 i5 8G | ubuntu 20.04 | 6m40 |
| macbookair 2020 i5 8G | macos catalina 10.15.7 | 9m50 |

### CODE REVIEW

File: docker-compose.full.yml [hunk 1]
```diff
@@ -159,6 +165,23 @@ services:
     labels:
       traefik.enable: "false"
 
+  node:
+    image: ${NODE_IMAGE:-node:14.15-alpine3.12}
```
Comment (comzeradd) [inline docker-compose.full.yml:169]: You could use `node:lts-alpine` to make sure we are always using the latest stable version.

Comment (lithrel) [reply docker-compose.full.yml:169]: Ah good point, I started with this one because it was the same version I saw locally; gonna try with lts :+1:

**Comment by comzeradd** [comment]
Once this is merged, we should remove the duplicate scripts from [base-fork](https://github.com/greenpeace/planet4-base-fork/tree/develop/tasks/other), right?

**Comment by lithrel** [comment]
> Once this is merged, we should remove the duplicate scripts from [base-fork](https://github.com/greenpeace/planet4-base-fork/tree/develop/tasks/other), right?

Yep that's the plan, I just need to be sure it's not used elsewhere

**Review by comzeradd — APPROVED** [review]
Tested it again, in a clean environment, after the latest changes, and still works great.

Not sure if you have anything else to add to this PR, but it looks good to be merged.

**Pull request merged**